Expanded  grpc calls

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Expanded the basenode grpc server to include the following calls:
Add get_block_template
Add get_block
Add get_meta_data
Add submit_block

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This allows a miner to connect the base_node via grpc. 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

philipr-za commented on applications/tari_base_node/src/grpc/server.rs:
```diff
@@ -296,6 +314,76 @@ impl base_node_grpc::base_node_server::BaseNode for BaseNodeGrpcServer {
+        let mut handler = self.node_service.clone();
+
+        let new_template = handler
+            .get_new_block_template(PowAlgorithm::Blake)
```
Shouldn't this be `algo`?

SWvheerden replied:
yip, thanks, fixed

StriderDM reviewed (dismissed):
Thanks for this! 

LGTM

philipr-za commented on applications/tari_base_node/src/grpc/server.rs:
```diff
@@ -666,7 +778,272 @@ impl From<BlockHeader> for base_node_grpc::BlockHeader {
+            total_kernel_offset: Vec::from(block.header.total_kernel_offset.as_bytes()),
+            pow: Some(base_node_grpc::ProofOfWork {
+                pow_algo: match block.header.pow.pow_algo {
+                    PowAlgorithm::Monero => 0,
```
I think you should rather implement a `From<PowAlgorithm> for u32` and a `TryFrom<u32> for PowAlgorithm` to handle these conversions. That way it is maintained in 1 clear place.

SWvheerden replied:
Wel thats currently used in a from, that means we have to change the entire conversion here to a try_from. 
At this point we dont have to check account for try's.

philipr-za commented on applications/tari_base_node/src/grpc/server.rs:
```diff
@@ -296,6 +314,76 @@ impl base_node_grpc::base_node_server::BaseNode for BaseNodeGrpcServer {
+    {
+        let request = request.into_inner();
+        debug!(target: LOG_TARGET, "Incoming GRPC request for get new block template");
+        let algo = match request.pow_algo {
```
I think you should rather implement a `From<PowAlgorithm> for u32` and a `TryFrom<u32> for PowAlgorithm` to handle these conversions. That way it is maintained in 1 clear place.

SWvheerden replied:
We have one, used that one

philipr-za commented on applications/tari_base_node/proto/base_node.proto:
```diff
@@ -223,6 +231,14 @@ message ProofOfWork {
+message PowAlgo{
+    // 0 = Monero
+    // 1 = Blake
+    uint64 pow_algo = 1;
```
Does this need to be a u64?

SWvheerden replied:
made smaller

stringhandler commented on applications/tari_base_node/src/grpc/server.rs:
```diff
@@ -666,7 +776,272 @@ impl From<BlockHeader> for base_node_grpc::BlockHeader {
+    }
+}
+
+impl From<NewBlockTemplate> for base_node_grpc::NewBlockTemplate {
```
I think put these conversions in a new file, or even separate files. This file is getting pretty big and these may be used elsewhere

stringhandler reviewed (dismissed):
Looks good. I think it'd be neater with the conversions moved into their own files, if you have time, but if not we can merge it as is.

SWvheerden committed:
Expanded grpc calls to allow miners to connect the base node

Add get_block_template
Add get_block
Add get_meta_data
Add submit_block
move conversions
